src/CBot/CBotInstr/CBotExprLitNum.cpp

```diff
@@ -75,6 +108,8 @@ CBotInstr* CBotExprLitNum::Compile(CBotToken* &p, CBotCStack* pStack)
 
         return pStack->Return(inst, pStk);
     }
+    // Program ended after the number. Return error indication.
+    // XXX: Why are we returning an error indication *here*, when the end-of-program doesn't actually affect the number parsing?
```

Comment:
In fact, the code after your comment will never execute because the next token will never be nullptr.

Reply:
NextToken returns true (and updates p by reference) if there was a next token and otherwise returns false. How do you know there will always be a next token?

Reply:
Currently the only places where numbers or other expressions will be compiled are within the definition of a function or class, which are verified by structure before any such expressions are compiled. Which is to say, always somewhere between curly braces or parentheses.

And beyond that, CBotToken::CompileTokens() appends a dummy token at the end, so the program will always have a first and last token.  The first token, by the way, is only used to consume any comments or white-space that may be at the top of the program script.

Reply:
I suppose this sort of thing is why the compiler needs a rewrite. (Does it though? Sometimes I jump into code and think it needs a rewrite, because I don't understand it yet)